### ISSUE TITLE
Readme.md

### ISSUE THREAD

**Issue by Data-pageup** [Data-pageup/delivery_time_optimizing_mlops_based#1]
# 📈 Time Series Analysis & Forecasting

A collection of lab assignments exploring fundamental and applied concepts in Time Series Analysis, implemented in Python using real and synthetic datasets.

**Student:** Amirtha Ganesh R  
**Reg No:** 24MSD7035  
**Course:** Time Series Analysis (MSD)

---

## 📂 Repository Structure

```
time-series-analysis/
│
├── Lab_01_Introduction_to_TSA/
│   └── lab1_electric_utilities.ipynb
│
├── Lab_02_White_Noise_and_Random_Walk/
│   └── lab2_white_noise_random_walk.ipynb
│
├── Lab_03_Exponential_Smoothing/
│   └── lab3_exponential_smoothing.ipynb
│
├── Lab_04_CO2_Forecasting/
│   └── lab4_co2_forecasting.ipynb
│
├── Lab_05_AR_Models_and_ARIMA/
│   └── lab5_ar3_arima.ipynb
│
└── README.md
```

---

### Lab 1 — Introduction to Time Series Analysis
**Dataset:** Industrial Production of Electric Utilities (IPG2211A2N), 1939–2019

- Data cleaning and dtype conversion
- Time series visualization
- Additive seasonal decomposition (trend, seasonal, residual components)
- Mean and variance computation at time `t` and `t+1`
- Autocovariance and autocorrelation at lags `t+1` and `t+12`
- ACF plot interpretation

**Key Finding:** The electric utilities production data is non-stationary — it shows a strong upward trend and seasonal fluctuations, confirmed by changing variance over time.

---

### Lab 2 — White Noise & Random Walk
**Dataset:** Synthetically generated (seed = 739 / 42)

- Generated 1,000 Gaussian random variables (μ=0, σ=1)
- Summary statistics and line plots
- ACF and PACF plots for white noise
- Generated a 1,000-step random walk via cumulative sum
- ADF (Augmented Dickey-Fuller) test for stationarity on both series

**Key Finding:**
| Property | White Noise | Random Walk |
|---|---|---|
| Mean | Stable (~0) | Drifts over time |
| Variance | Constant | Increases with time |
| Stationary? | ✅ Yes (ADF p≈0.0) | ❌ No (ADF p≈0.77) |

---

### Lab 3 — Exponential Smoothing
**Dataset:** Synthetic data — stationary, trending, and seasonal signals

Implemented and compared three smoothing methods across multiple parameter settings:

- **Simple Exponential Smoothing (SES):** 5 models with α ∈ {0.1, 0.3, 0.5, 0.7, 0.9}
- **Double Exponential Smoothing (DES / Holt's):** 5 models with varying α and β
- **Triple Exponential Smoothing (TES / Holt-Winters):** 5 models with α, β, and γ combinations

**Key Finding:**
- SES best suits stationary data; α ≈ 0.3 gives the best balance of smoothing vs. responsiveness
- DES captures linear trends; lower α and β values yield more stable forecasts
- TES handles seasonality; high values for all three parameters lead to overfitting and instability

---

### Lab 4 — CO2 Forecasting with Exponential Smoothing
**Dataset:** Statsmodels built-in CO2 dataset (Mauna Loa, 1958–2001), resampled monthly

- Loaded, resampled to monthly frequency, and forward/backward filled missing values
- Visualized monthly CO2 concentration and performed additive decomposition
- Train/test split: last 48 months held out as test set
- Trained SES, DES, and TES models and evaluated on test set

**Model Comparison:**

| Model | MAE | MSE |
|---|---|---|
| Single ES | 5.78 | 39.07 |
| Double ES | 11.02 | 143.81 |
| **Triple ES** | **0.93** | **1.14** |

- Forecast of next 6 months (Jan–Jun 2002) generated using the best-performing Triple ES model

**Key Finding:** Triple Exponential Smoothing significantly outperforms the other models because the CO2 data exhibits both a clear trend and annual seasonality.

---

### Lab 5 — AR(3) Models, Yule-Walker & ARIMA
**Dataset:** Synthetically generated zero-mean AR(3) series (seed = 666, n = 500)

- Simulated AR(3) process: `X_t = 0.1·X_{t-1} + 0.1·X_{t-2} + 0.1·X_{t-3} + ε_t`
- Plotted the series and its ACF/PACF
- ADF test confirmed stationarity (p ≈ 0.0)
- Estimated AR(3) parameters using the **Yule-Walker equations** (manual matrix solve)
- Computed and validated residuals for white noise (Ljung-Box test) and normality (Anderson-Darling test)
- Fit `ARIMA(3,0,0)` using `statsmodels` and compared parameter estimates
- Forecasted X₅₀₁, X₅₀₂, X₅₀₃ using both methods

**Estimated AR Coefficients:**

| Method | φ₁ | φ₂ | φ₃ |
|---|---|---|---|
| Yule-Walker | 0.0422 | 0.1248 | 0.1239 |
| ARIMA | 0.0423 | 0.1248 | 0.1246 |

**Key Finding:** Both methods yield nearly identical parameter estimates. Residuals passed both the white noise and normality checks, validating the model fit.

---

## 🛠️ Tech Stack

- **Python 3.x**
- `pandas`, `numpy` — data manipulation
- `matplotlib`, `seaborn` — visualization
- `statsmodels` — decomposition, ACF/PACF, ADF test, ARIMA, Exponential Smoothing
- `scikit-learn` — MAE/MSE evaluation
- `scipy` — Anderson-Darling normality test

---

## ▶️ Getting Started

```bash
# Clone the repository
git clone https://github.com/<your-username>/time-series-analysis.git
cd time-series-analysis

# Install dependencies
pip install pandas numpy matplotlib seaborn statsmodels scikit-learn scipy

# Launch Jupyter
jupyter notebook
```

---

## 📚 Concepts Covered

- Time series components: trend, seasonality, residuals
- Stationarity and unit root tests (ADF)
- White noise and random walks
- Autocovariance, autocorrelation (ACF), partial autocorrelation (PACF)
- Simple, Double, and Triple Exponential Smoothing (Holt-Winters)
- Autoregressive (AR) models and Yule-Walker estimation
- ARIMA modeling and forecasting

---

## 📝 License

This repository is intended for educational purposes.
